File: WeatherForecast/WeatherForecast/ViewControllers/WeatherDetailVC.m
//
//  CityDetailVC.m
//  WeatherForecast
//
//  Created by Hakkı Yiğit Yener on 23.05.2020.
//  Copyright © 2020 N11.com. All rights reserved.
//

#import "WeatherDetailVC.h"
#import "NetworkManager.h"
#import "WeatherDailyForecastListModel.h"
#import "TitleValueModel.h"
#import "WeatherForecast-Swift.h"


@interface WeatherDetailVC ()<UITableViewDelegate,UITableViewDataSource>
@property (weak, nonatomic) IBOutlet UITableView *weatherTableView;
@property (nonatomic, strong) NSArray* dataSource;
@property (nonatomic, strong) WeatherDailyForecastResponseModel* response;

@end

@implementation WeatherDetailVC

- (void)viewDidLoad {
    [super viewDidLoad];
    self.title =  [NSString stringWithFormat:@"%@ Hava Durumu",self.city.name];
    [self registerCells];
    [self loadWeatherDetail];
    
}

-(void)registerCells {
    [self.weatherTableView registerNib:[UINib nibWithNibName:@"WeatherDetailCell" bundle:nil]
               forCellReuseIdentifier:@"WeatherDetailCell"];
}

-(void)loadWeatherDetail {
    [self showLoading];
    [[NetworkManager shared] getWeatherDetail:self.city
                            complationHandler:^(BOOL success, WeatherDailyForecastResponseModel * _Nullable responseData, NSError * _Nullable error) {
        [self hideLoading];
        if (success && !error) {
            self.response = responseData;
            [self createDataSource:responseData];
        }
        else {
            [self showAlertWithTitle:@"Hata"
                             message:error.localizedDescription];
        }
    }];
}

-(void)createDataSource:(WeatherDailyForecastResponseModel*)response {
    NSMutableArray* daysArray = [[NSMutableArray alloc] init];
    for (WeatherDailyForecastListModel* item in response.list) {
        [daysArray addObject:[item convertToPropertyArray]];
    }
    self.dataSource = daysArray;
    [self.weatherTableView reloadData];
}
#pragma mark - UITableView Methods
-(NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    WeatherDailyForecastListModel* dailyWeather = [self.response.list objectAtIndex:section];
    
    NSDateFormatter* df = [[NSDateFormatter alloc]init];
    [df setDateFormat:@"dd-MM-yyyy"];
    return [df stringFromDate:dailyWeather.dt];
}

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    WeatherDetailCellSwift* cell =  [tableView dequeueReusableCellWithIdentifier:@"WeatherDetailCell"
                                                      forIndexPath:indexPath];
    NSArray* dayArray = [self.dataSource objectAtIndex:indexPath.section];
    TitleValueModel* rowValue = [dayArray objectAtIndex:indexPath.row];
    [cell refreshCellWith:rowValue];
    return cell;
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    NSArray* dayArray = [self.dataSource objectAtIndex:section];
    return dayArray.count;
}
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return self.dataSource.count;
}

-(CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 44;
}

@end

File: WeatherForecast/WeatherForecast/ViewControllers/WeatherDetailVC.swift
//
//  WeatherDetailVC.swift
//  WeatherForecast
//
//  Created by Hakkı Yiğit Yener on 27.05.2020.
//  Copyright © 2020 N11.com. All rights reserved.
//

import Foundation


class WeatherDetailVC: UIViewController, UITableViewDelegate, UITableViewDataSource {
    @IBOutlet private weak var weatherTableView: UITableView!
    @objc var city: City?
    private var dataSource: [[TitleValueModel]]?
    private var response: WeatherDailyForecastResponseModel?

    override func viewDidLoad() {
        super.viewDidLoad()
        title = "\(city?.name ?? "") Hava Durumu"
        registerCells()
        loadWeatherDetail()

    }

    func registerCells() {
        weatherTableView.register(
            UINib(nibName: "WeatherDetailCell", bundle: nil),
            forCellReuseIdentifier: "WeatherDetailCell")
    }
    func loadWeatherDetail() {
        showLoading()
        NetworkManager.shared.getWeatherDetail(
            city,
            complationHandler: { success, responseData, error in
                self.hideLoading()
                if success && error == nil {
                    self.response = responseData
                    self.createDataSource(responseData)
                } else {
                    self.showAlert(
                        withTitle: "Hata",
                        message: error?.localizedDescription)
                }
            })
    }
    func createDataSource(_ response: WeatherDailyForecastResponseModel?) {
        var daysArray: [[TitleValueModel]] = []
        if let list = response?.list {
            for item in list {
                guard let item = item as? WeatherDailyForecastListModel else {
                    continue
                }
                //        [daysArray addObject:[item convertToPropertyArray]];
            }
        }
        dataSource = daysArray
        weatherTableView.reloadData()
    }
// MARK: - UITableView Methods
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        let dailyWeather = response?.list[section]

        let df = DateFormatter()
        df.dateFormat = "dd-MM-yyyy"
        return "Today" //[df stringFromDate:dailyWeather.dt];
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(
            withIdentifier: "WeatherDetailCell",
            for: indexPath) as? WeatherDetailCellSwift
        let dayArray = dataSource?[indexPath.section]
        let rowValue = dayArray?[indexPath.row] as? TitleValueModel
        cell?.refreshCell(with: rowValue)
        return cell!
    }
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let dayArray = dataSource?[section]
        return dayArray?.count ?? 0
    }

    func numberOfSections(in tableView: UITableView) -> Int {
        return dataSource?.count ?? 0
    }

    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 44
    }

}
